src/gpu/nvidia/cudnn_matmul_lt_impl.hpp

```diff
@@ -0,0 +1,726 @@
+        cublasLtReductionScheme_t reduction_scheme
+                = CUBLASLT_REDUCTION_SCHEME_MASK;
+        CUBLAS_EXECUTE_FUNC(cublasLtMatmulPreferenceSetAttribute, preference_,
+                CUBLASLT_MATMUL_PREF_REDUCTION_SCHEME_MASK, &reduction_scheme,
+                sizeof(reduction_scheme));
+
+        int num_results = 0;
+        if (imma_ampere_case_) {
+            CUBLAS_EXECUTE_FUNC(cublasLtMatmulAlgoGetHeuristic, lt_handle,
+                    operation_desc_, blocked_a_layout_, blocked_b_layout_,
+                    blocked_c_layout_, blocked_c_layout_, preference_,
+                    1 /* Num requested algos*/, &heuristic_results_,
+                    &num_results);
+        } else {
+            CUBLAS_EXECUTE_FUNC(cublasLtMatmulAlgoGetHeuristic, lt_handle,
+                    operation_desc_, a_layout_, b_layout_, c_layout_, c_layout_,
+                    preference_, 1 /* Num requested algos*/,
+                    &heuristic_results_, &num_results);
+        }
+
+        if (num_results == 0) { return status_t::dnnl_runtime_error; }
+        gemm_algo_ = heuristic_results_.algo;
+        algo_scratch_size_ = heuristic_results_.workspaceSize;
+        return status_t::dnnl_success;
+    }
+
+    void init_scratchpad(matmul_pd_t *pd) override {
+        auto scratchpad = pd->scratchpad_registry().registrar();
+        if (algo_scratch_size_ > 0) {
+            scratchpad.book(memory_tracking::names::key_matmul_lt_algo_scratch,
+                    algo_scratch_size_, 1, 256);
+        }
+        if (weight_size_ > 0) {
+            scratchpad.book(memory_tracking::names::key_matmul_lt_block_a,
+                    weight_size_, 1, 256);
+        }
+        if (source_size_ > 0) {
+            scratchpad.book(memory_tracking::names::key_matmul_lt_block_b,
+                    source_size_, 1, 256);
+        }
+        if (dest_size_ > 0) {
+            scratchpad.book(memory_tracking::names::key_matmul_lt_block_c,
+                    dest_size_, 1, 256);
+        }
+    }
+
+    void execute(cublasHandle_t cublas_handle, cudnnHandle_t cudnn_handle,
+            void *a, void *b, void *c, void *bias, void *algo_scratch,
+            void *reorder_scratch, void *block_a_scratch, void *block_b_scratch,
+            void *block_c_scratch, void *src_scale, void *wei_scale,
+            void *dst_scale) override {
+
+        cudaStream_t streamId;
+        auto lt_handle = (cublasLtHandle_t)(cublas_handle);
+        CUBLAS_EXECUTE_FUNC(cublasGetStream, cublas_handle, &streamId);
+
+        if (imma_ampere_case_) {
+            transform_matrix(lt_handle, b_layout_, b, blocked_b_layout_,
+                    block_b_scratch, !trans_b_, streamId);
+            b = block_b_scratch;
+            if (!w_blocked_) {
+                transform_matrix(lt_handle, a_layout_, a, blocked_a_layout_,
+                        block_a_scratch, trans_a_, streamId);
+                a = block_a_scratch;
+            }
```

Comment:
I guess the most optimal would be to do the reorders separate from execute call, to let user schedule those and reduce copy overheads.

In any case, is there any benefit to do this reorder inside execute call?

Reply:
Without adding this transform inside the excecute the IMMA kernels for the cublasLT will only run with `Ab32a` for the weights and output with the transform we can support other data formats as well.

The user can do the transform currently for the weights by reordering to `Ab32a` and passing it to the matmul pd as such, the transform will not be called in this case.

For now we do not have a reorder format mapping to the input matrix to the ampere blocked format (due to the interleaved nature of the blocking) hence we need to do the transform inside the execute.

For the output to schedule a transform at another time the user can set the output of the matmul to `Ab32a` and reorder after.